concurrency/LocalMUCRoom.java

```diff
@@ -0,0 +1,3450 @@
+// *** END amazon ***
+            for (MUCRole occupant : occupantsByFullJID.values()) {
+                if (occupant == joinRole) {
+                    continue;
+                }
+                Presence occupantPresence = occupant.getPresence();
+                // Skip to the next occupant if we cannot send presence of this occupant
+                if (hasToCheckRoleToBroadcastPresence()) {
+                    Element frag = occupantPresence.getChildElement("x",
+                            "http://jabber.org/protocol/muc#user");
+                    // Check if we can broadcast the presence for this role
+                    if (!canBroadcastPresence(frag.element("item").attributeValue("role"))) {
+                        continue;
+                    }
+                }
+                // Don't include the occupant's JID if the room is semi-anon and the new occupant
+                // is not a moderator
+                if (!canAnyoneDiscoverJID() && MUCRole.Role.moderator != joinRole.getRole()) {
+                    occupantPresence = occupantPresence.createCopy();
+                    Element frag = occupantPresence.getChildElement("x",
+                            "http://jabber.org/protocol/muc#user");
+                    frag.element("item").addAttribute("jid", null);
+                }
+                joinRole.send(occupantPresence);
+            }
+// *** BEGIN amazon ***
+        }
+// *** END amazon ***
+    }
+
+    public void occupantAdded(OccupantAddedEvent event) {
+        // Create a proxy for the occupant that joined the room from another cluster node
+        RemoteMUCRole joinRole = new RemoteMUCRole(mucService, event);
+        JID bareJID = event.getUserAddress().asBareJID();
+        String nickname = event.getNickname();
+        List<MUCRole> occupants = occupantsByNickname.get(nickname.toLowerCase());
+        // Do not add new occupant with one with same nickname already exists
+        if (occupants == null) {
+            occupants = new CopyOnWriteArrayList<>(); // *** ONE_LINE amazon
+            occupantsByNickname.put(nickname.toLowerCase(), occupants);
+        } else {
+            // sanity check; make sure the nickname is owned by the same JID
+            if (occupants.size() > 0) {
+                JID existingJID = occupants.get(0).getUserAddress().asBareJID();
+                if (!bareJID.equals(existingJID)) {
+                    Log.warn(MessageFormat.format("Conflict detected; {0} requested nickname '{1}'; already being used by {2}", bareJID, nickname, existingJID));
+                    return;
+                }
+            }
+        }
+        // Add the new user as an occupant of this room
+        occupants.add(joinRole);
+        // Update the tables of occupants based on the bare and full JID
+        List<MUCRole> list = occupantsByBareJID.get(bareJID);
+        if (list == null) {
+            list = new ArrayList<>();
+            occupantsByBareJID.put(bareJID, list);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **951, 952, and 954**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and uses that reference at line: **956**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.